src/cgame/cg_rocket_draw.cpp

```diff
@@ -2174,6 +2174,131 @@ class PredictedMineEfficiencyElement : public HudElement
 	std::unordered_map<int, std::string> pluralSuffix;
 };
 
+class BarbsHudElement : public HudElement
+{
+public:
+	BarbsHudElement ( const Rocket::Core::String& tag ) :
+	HudElement ( tag, ELEMENT_ALIENS ),
+	numBarbs( 0 ),
+	maxBarbs( 0 ),
+	regenerationInterval ( 0 ),
+	t0 ( 0 ),
+	offset ( 0 ) {}
+
+	void OnAttributeChange( const Rocket::Core::AttributeNameList& changed_attributes )
+	{
+		HudElement::OnAttributeChange( changed_attributes );
+		if ( changed_attributes.find( "src" ) != changed_attributes.end() )
+		{
+			maxBarbs = BG_Weapon( WP_ALEVEL3_UPG )->maxAmmo;
```

Comment:
As it stands the data accessed through `BG_Weapon` isn't going to change within a match. As long as new `BarbsHudElement` objects are created on each `vid_restart` or map load (@DolceTriade, is that the case?), then you could also add this to the initialization list above. In direct comparisong it looks like your current approach is more error-resistant but the assumption that the value of `BG_Weapon( WP_ALEVEL3_UPG )->maxAmmo` can change between calls to `OnAttributeChange` but not between calls to `DoOnUpdate` seems a bit arbitrary to me.